readme.md
# randomsp
> A library to pull and return a random stock from a smattering of stock indices

Based on some research through computer simulations, it was found that perhaps a more random strategy to investing in the stock market was more stable over the long term than using any particular strategy.

[Link to forbes article that explains the study](https://www.forbes.com/sites/alexknapp/2013/03/22/computer-simulation-suggests-that-the-best-investment-strategy-is-a-random-one/#2189846a5136)

In the same vein of thought, randomsp was born.

## Indices
The study from the article used 5 different stock indices to make their predictions: the UK FTSE, the MIB FTSE (Italian stock exchange), the Nikkei 225 (Japanese market), the DAX (German market), and the S&P 500.

This library includes those 5 indices plus the Nasdaq stock exchange.

## Notes
- The Nikkei 225 stocks have symbols, but more often than not are denoted by a number value in their index.  Because of this, the name of the company and number are returned instead of the symbol acryonym.

## Install
```
$ go get github.com/ru-lai/randomsp
```

## Usage
```
import "github.com/ru-lai/randomsp"

func main() {
	stock, err := randomsp.GetRandomIndexStock()
	// ZTS

	stock, err := randomsp.GetRandomNasdaqStock()
	// ATVI

	stock, err := randomsp.GetRandomFinancialTimesStock()
	// G4S

	stock, err := randomsp.GetRandomSPStock()
	// WRK

	stock, err := randomsp.GetRandomItalianFinancialTimesStock()
	// BPE

        stock, err := randomsp.GetRandomNikkeiStock()
        // Tokyu Land 3289

	stock, err := randomsp.GetRandomDaxStock()
	// BMW

        // each function call should have error handling
        if err != nil {
            // handle error here
            log.Fatal(err)
        }
}
```

## API
Each method will return a string of a random stock from either the index you called or a random index and an error if applicable

## License

MIT © Tyler Boright


changelog.md
# RandomSP

## [Unreleased]

## [1.2.2] - 2020-06-10
- Adds changelog to the repo and fixes import issue.

## [1.2.1] - 2020-06-10
- Fixes bug with the S&P500 stock symbols where no symbol is returned.


randomsp_test.go
package randomsp

import (
	"math/rand"
	"regexp"
	"strings"
	"testing"
	"time"
)

func TestUnitGetRandomString(t *testing.T) {
	rand.Seed(time.Now().UnixNano())
	str := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	ss := strings.Split(str, "")

	if getRandomString(ss) == getRandomString(ss) {
		t.Error("The Get Rand String function should get different strings from the slice")
	}
}

type testStockIndex struct {
	IndexName string
	Stocks    []string
}

// a test to hit all stock data sources and verify that acronyms are returned
func TestIntegrationAcronymsReturned(t *testing.T) {
	var stocks []testStockIndex

	dax, err := getDaxStocks()
	if err != nil {
		t.Fatalf("acronym integration test failed getting Dax stocks with the following error: %s", err)
	}

	daxIdx := testStockIndex{
		"DAX",
		dax,
	}

	ft, err := getFinancialTimesStocks()
	if err != nil {
		t.Fatalf("acronym integration test failed getting Financial Times stocks with the following error: %s", err)
	}

	ftIdx := testStockIndex{
		"FinancialTimes",
		ft,
	}

	ift, err := getItalianFinancialTimesStocks()
	if err != nil {
		t.Fatalf("acronym integration test failed getting Italian Financial Times stocks with the following error: %s", err)
	}

	iftIdx := testStockIndex{
		"ItalianFinancialTimes",
		ift,
	}

	n, err := getNasdaqStocks()
	if err != nil {
		t.Fatalf("acronym integration test failed getting Nasdaq stocks with the following error: %s", err)
	}

	nIdx := testStockIndex{
		"Nasdaq",
		n,
	}

	nik, err := getNikkeiStocks()
	if err != nil {
		t.Fatalf("acronym integration test failed getting Nikkei stocks with the following error: %s", err)
	}

	nikIdx := testStockIndex{
		"Nikkei",
		nik,
	}

	sp, err := getStandardPoorsStocks()
	if err != nil {
		t.Fatalf("acronym integration test failed getting StandardPoors stocks with the following error: %s", err)
	}

	spIdx := testStockIndex{
		"StandardPoor",
		sp,
	}

	stocks = append(stocks, daxIdx)
	stocks = append(stocks, ftIdx)
	stocks = append(stocks, iftIdx)
	stocks = append(stocks, nIdx)
	stocks = append(stocks, nikIdx)
	stocks = append(stocks, spIdx)

	// Create Regex To Test that inputs have sequences of capital letters (prove that it is a stock quote acryonym
	stockRegex := regexp.MustCompile("[A-Z]+")

	for _, idxSlice := range stocks {
		for _, stock := range idxSlice.Stocks {
			if stockRegex.FindStringIndex(stock) == nil {
				t.Fatalf("the stock returned does not match the stock regex and is probably not a stock.  \nIndex: %q\nStock: %q", idxSlice.IndexName, stock)
			}
		}
	}
}

// create a struct to allow easier printing of the function name
type testStockFn struct {
	Name string
	Fn   func() (Stock, error)
}

func TestIntegrationStockStringsRemoveSpaces(t *testing.T) {
	stockFuncs := []testStockFn{
		testStockFn{Name: "GetRandomNasdaqStock", Fn: GetRandomNasdaqStock},
		testStockFn{Name: "GetRandomNikkeiStock", Fn: GetRandomNikkeiStock},
		testStockFn{Name: "GetRandomSPStock", Fn: GetRandomSPStock},
		testStockFn{Name: "GetRandomFinancialTimesStock", Fn: GetRandomFinancialTimesStock},
		testStockFn{Name: "GetRandomItalianFinancialTimesStock", Fn: GetRandomItalianFinancialTimesStock},
		testStockFn{Name: "GetRandomDaxStock", Fn: GetRandomDaxStock},
	}

	for _, stockFn := range stockFuncs {
		stock, err := stockFn.Fn()
		if err != nil {
			t.Fatalf("The stockFn did not return a stock correctly. Stock Function: %s\nError: %s\n", stockFn.Name, err)
		}

		sLen := len(stock.Symbol)
		if string(stock.Symbol[0]) == " " || string(stock.Symbol[sLen-1]) == " " {
			t.Fatalf("A stock was returned with an untrimmed Symbol. Symbol: %q\n, First Character: %s\n, Last Character: %s\n", stock.Symbol, string(stock.Symbol[0]), string(stock.Symbol[sLen-1]))
		}
	}
}


main.go
package main

import (
	"github.com/PuerkitoBio/goquery"
	"log"
	"math/rand"
	"net/http"
	"time"
)

func GetRandomInt(min, max int) int {
	return min + rand.Intn(max-min)
}

func GetSPStocks() (stocks []string) {
	res, err := http.Get("https://en.wikipedia.org/wiki/List_of_S%26P_500_companies")
	if err != nil {
		log.Fatal(err)
	}

	defer res.Body.Close()

	if res.StatusCode != 200 {
		log.Fatalf("Status code error: %d %s", res.StatusCode, res.Status)
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		log.Fatal(err)
	}

	tbody := doc.Find("tbody").First()
	tbody.Find("tr").Each(func(i int, s *goquery.Selection) {
		td := s.Find("td:first-child")
		stocks = append(stocks, td.Text())
	})
	return
}

func GetRandString(ss []string) string {
	randInt := GetRandomInt(1, len(ss))
	return ss[randInt:(randInt + 1)][0]
}

func main() {
	rand.Seed(time.Now().UnixNano())
	stockSlice := GetSPStocks()
	// log.Printf("%+v", stockSlice)
	log.Println(GetRandString(stockSlice))
	for i := 0; i < 1000; i++ {
		log.Println(GetRandString(stockSlice))
	}
}


randomsp.go
package randomsp

import (
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"github.com/tlboright/go-rint"
	"net/http"
	"strings"
)

func getDaxStocks() (stocks []string, err error) {
	res, err := http.Get("https://en.wikipedia.org/wiki/DAX")
	if err != nil {
		return
	}

	defer res.Body.Close()

	if res.StatusCode != 200 {
		return
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return
	}

	tables := doc.Find(".mw-parser-output > table")
	tables.Filter("#constituents").Find("tr").Each(func(i int, s *goquery.Selection) {
		td := s.Find("td:nth-child(4)")
		if td.Text() != "" {
			stocks = append(stocks, td.Text())
		}
	})
	return
}

func getFinancialTimesStocks() (stocks []string, err error) {
	res, err := http.Get("https://en.wikipedia.org/wiki/FTSE_100_Index")
	if err != nil {
		return
	}

	defer res.Body.Close()

	if res.StatusCode != 200 {
		return
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return
	}

	tbody := doc.Find("#constituents > tbody")
	tbody.Find("tr").Each(func(i int, s *goquery.Selection) {
		td := s.Find("td:nth-child(2)")
		if td.Text() != "" {
			stocks = append(stocks, td.Text())
		}
	})
	return
}

func getItalianFinancialTimesStocks() (stocks []string, err error) {
	res, err := http.Get("https://en.wikipedia.org/wiki/FTSE_MIB")
	if err != nil {
		return
	}

	defer res.Body.Close()

	if res.StatusCode != 200 {
		return
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return
	}

	tbody := doc.Find("#constituents > tbody")
	tbody.Find("tr").Each(func(i int, s *goquery.Selection) {
		td := s.Find("td:nth-child(2)")
		if td.Text() != "" {
			stocks = append(stocks, td.Text())
		}
	})
	return
}

func getNasdaqStocks() (stocks []string, err error) {
	res, err := http.Get("https://en.wikipedia.org/wiki/NASDAQ-100")
	if err != nil {
		return
	}

	defer res.Body.Close()

	if res.StatusCode != 200 {
		return
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return
	}

	tbody := doc.Find("#constituents > tbody")
	tbody.Find("tr").Each(func(i int, s *goquery.Selection) {
		td := s.Find("td:nth-child(2)")
		if td.Text() != "" {
			stocks = append(stocks, td.Text())
		}
	})
	return
}

func getNikkeiStocks() (stocks []string, err error) {
	res, err := http.Get("https://en.wikipedia.org/wiki/Nikkei_225")
	if err != nil {
		return
	}

	defer res.Body.Close()

	if res.StatusCode != 200 {
		return
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return
	}

	// the list of stock names underneath the titles
	doc.Find("h3 + ul").Each(func(i int, s *goquery.Selection) {
		lis := s.Find("li")
		lis.Each(func(i int, s *goquery.Selection) {
			name := s.Find("a:nth-child(1)")
			symbol := s.Find("a:nth-child(3)")

			if name.Text() != "" && symbol.Text() != "" {
				stocks = append(stocks, fmt.Sprintf("%s %s\n", name.Text(), symbol.Text()))
			}
		})
	})

	return
}

func getStandardPoorsStocks() (stocks []string, err error) {
	res, err := http.Get("https://en.wikipedia.org/wiki/List_of_S%26P_500_companies")
	if err != nil {
		return
	}

	defer res.Body.Close()

	if res.StatusCode != 200 {
		return
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return
	}

	tbody := doc.Find("#constituents > tbody")
	tbody.Find("tr").Each(func(i int, s *goquery.Selection) {
		td := s.Find("td:first-child")
		if td.Text() != "" {
			stocks = append(stocks, td.Text())
		}
	})
	return
}

func getRandomString(ss []string) string {
	randInt := rint.GenRange(1, len(ss))
	return ss[randInt:(randInt + 1)][0]
}

func GetRandomDaxStock() (stock Stock, err error) {
	rint.Init()
	stockSlice, err := getDaxStocks()
	if err != nil {
		return
	}

	stock = Stock{strings.TrimSpace(getRandomString(stockSlice)), "Dax"}
	return
}

func GetRandomFinancialTimesStock() (stock Stock, err error) {
	rint.Init()
	stockSlice, err := getFinancialTimesStocks()
	if err != nil {
		return
	}

	stock = Stock{strings.TrimSpace(getRandomString(stockSlice)), "Financial Times"}
	return
}

func GetRandomItalianFinancialTimesStock() (stock Stock, err error) {
	rint.Init()
	stockSlice, err := getItalianFinancialTimesStocks()
	if err != nil {
		return
	}

	stock = Stock{strings.TrimSpace(getRandomString(stockSlice)), "Italian Financial Times"}
	return
}

func GetRandomNasdaqStock() (stock Stock, err error) {
	rint.Init()
	stockSlice, err := getNasdaqStocks()
	if err != nil {
		return
	}

	stock = Stock{strings.TrimSpace(getRandomString(stockSlice)), "Nasdaq"}

	return
}

func GetRandomNikkeiStock() (stock Stock, err error) {
	rint.Init()
	stockSlice, err := getNikkeiStocks()
	if err != nil {
		return
	}

	stock = Stock{strings.TrimSpace(getRandomString(stockSlice)), "Nikkei"}

	return
}

func GetRandomSPStock() (stock Stock, err error) {
	rint.Init()
	stockSlice, err := getStandardPoorsStocks()
	if err != nil {
		return
	}

	stock = Stock{strings.TrimSpace(getRandomString(stockSlice)), "S&P 500"}

	return
}

func GetRandomIndexStock() (stock Stock, err error) {
	rint.Init()
	stockFuncs := []func() (Stock, error){GetRandomNasdaqStock, GetRandomSPStock, GetRandomNikkeiStock, GetRandomFinancialTimesStock, GetRandomItalianFinancialTimesStock, GetRandomDaxStock}
	stock, err = stockFuncs[rint.GenRange(0, len(stockFuncs))]()
	return
}
